File: frontend/src/api/pilotModelFlightTime.ts
import request from '@/utils/request';
import type {
    PilotModelFlightTime,
    CreatePilotModelFlightTimeDto,
    UpdatePilotModelFlightTimeDto,
    ListQuery,
    PaginatedResponse,
} from '@/types/pilotModelFlightTime';

const BASE_URL = '/pilot-model-flight-time';

/** 新增 */
export function createPilotModelFlightTime(
    data: CreatePilotModelFlightTimeDto
): Promise<PilotModelFlightTime> {
    return request.post(BASE_URL, data);
}

/** 根据 ID 查询 */
export function getPilotModelFlightTimeById(
    id: number
): Promise<PilotModelFlightTime> {
    return request.get(`${BASE_URL}/${id}`);
}

/** 根据 pilot_id + model_id 查询 */
export function getByPilotAndModel(
    pilot_id: number,
    model_id: number
): Promise<PilotModelFlightTime> {
    return request.get(`${BASE_URL}/pilot/${pilot_id}/model/${model_id}`);
}

/** 列表查询（支持分页和过滤） */
export function listPilotModelFlightTimes(
    params: { pilot_id?: number; model_id?: number }
): Promise<{ list: PilotModelFlightTime[] }> {
    return request.get(`${BASE_URL}/list`, { params });
}

/** 更新（根据 ID） */
export function updatePilotModelFlightTime(
    id: number,
    data: UpdatePilotModelFlightTimeDto
): Promise<PilotModelFlightTime> {
    return request.put(`${BASE_URL}/${id}`, data);
}

/** Upsert（存在则累加并更新时间，否则新增） */
export function upsertPilotModelFlightTime(
    data: CreatePilotModelFlightTimeDto
): Promise<PilotModelFlightTime> {
    return request.post(`${BASE_URL}/upsert`, data);
}

/** 根据 ID 删除 */
export function deletePilotModelFlightTime(id: number): Promise<void> {
    return request.delete(`${BASE_URL}/${id}`);
}

/** 根据 pilot_id 批量删除 */
export function deleteByPilot(pilot_id: number): Promise<{ deleted: number }> {
    return request.delete(`${BASE_URL}/pilot/${pilot_id}`);
}

File: backend/src/controller/pilot_model_flight_time.rs
use actix_web::{web, HttpResponse};
use crate::error::AppResult;
use crate::service::ServiceContainer;
use crate::service::pilot_model_flight_time::{CreatePilotModelFlightTimeDto, UpdatePilotModelFlightTimeDto, ListQuery};
use crate::utils::response::success_response;

/// 新增
pub async fn create(
    container: web::Data<ServiceContainer>,
    dto: web::Json<CreatePilotModelFlightTimeDto>,
) -> AppResult<HttpResponse> {
    let record = container.pilot_model_flight_time_service.create(dto.into_inner()).await?;
    Ok(success_response(record))
}

/// 根据 ID 查询
pub async fn get_by_id(
    container: web::Data<ServiceContainer>,
    path: web::Path<i64>,
) -> AppResult<HttpResponse> {
    let id = path.into_inner();
    let record = container.pilot_model_flight_time_service.get_by_id(id).await?
        .ok_or_else(|| crate::error::AppError::RecordNotFound)?;
    Ok(success_response(record))
}

/// 根据 pilot_id + model_id 查询
pub async fn get_by_pilot_model(
    container: web::Data<ServiceContainer>,
    path: web::Path<(i64, i64)>,
) -> AppResult<HttpResponse> {
    let (pilot_id, model_id) = path.into_inner();
    let record = container.pilot_model_flight_time_service
        .get_by_pilot_and_model(pilot_id, model_id)
        .await?
        .ok_or_else(|| crate::error::AppError::RecordNotFound)?;
    Ok(success_response(record))
}

/// 列表查询
pub async fn list(
    container: web::Data<ServiceContainer>,
    query: web::Query<ListQuery>,
) -> AppResult<HttpResponse> {
    let (items) = container.pilot_model_flight_time_service.list(query.into_inner()).await?;
    Ok(success_response(serde_json::json!({
        "list": items,
    })))
}

/// 更新（根据 ID）
pub async fn update(
    container: web::Data<ServiceContainer>,
    path: web::Path<i64>,
    dto: web::Json<UpdatePilotModelFlightTimeDto>,
) -> AppResult<HttpResponse> {
    let id = path.into_inner();
    let updated = container.pilot_model_flight_time_service.update(id, dto.into_inner()).await?;
    Ok(success_response(updated))
}

/// Upsert
pub async fn upsert(
    container: web::Data<ServiceContainer>,
    dto: web::Json<CreatePilotModelFlightTimeDto>,
) -> AppResult<HttpResponse> {
    let record = container.pilot_model_flight_time_service.upsert(dto.into_inner()).await?;
    Ok(success_response(record))
}

/// 根据 ID 删除
pub async fn delete(
    container: web::Data<ServiceContainer>,
    path: web::Path<i64>,
) -> AppResult<HttpResponse> {
    let id = path.into_inner();
    container.pilot_model_flight_time_service.delete(id).await?;
    Ok(success_response(()))
}

/// 根据 pilot_id 批量删除
pub async fn delete_by_pilot(
    container: web::Data<ServiceContainer>,
    path: web::Path<i64>,
) -> AppResult<HttpResponse> {
    let pilot_id = path.into_inner();
    let affected = container.pilot_model_flight_time_service.delete_by_pilot(pilot_id).await?;
    Ok(success_response(serde_json::json!({ "deleted": affected })))
}